x/mongo/driver/topology/connection.go

```diff
@@ -108,7 +109,19 @@ func (c *connection) connect(ctx context.Context) {
 	}
 	defer close(c.connectDone)
 
+	c.mu.Lock()
 	ctx, c.cancelConnectContext = context.WithCancel(ctx)
+	c.mu.Unlock()
+
+	defer func() {
+		c.mu.Lock()
+		if c.cancelConnectContext != nil {
+			c.cancelConnectContext()
```

Comment:
We should look into the cost of context cancellation because I don't know how expensive this function call is. If it turns out to be pretty expensive, we should avoid doing it in lock:

```
var cancelFn context.CancelFunc

c.mu.Lock()
cancelFn = c.cancelConnectContext
c.cancelConnectContext = nil
c.mu.Unlock()

if cancelFn != nil {
    cancelFn()
}
```

A similar change could be made to the other function as well.

Reply:
Looks like context cancellation is relatively expensive, I've made this change